STD-40 | Realize TheoryBlock Controller

zefirlover opened the pull request.

Evenya committed:
Query Task added

Evenya committed:
created: // AddTheoryBlockCommand.cs
// GetTheoryBlockByIdQuery.cs
// RemoveTheoryBlockByIdCommand.cs

maridodonova committed:
Added course controller and course requests and responses

maridodonova committed:
Fixed http get/post

maridodonova committed:
Added course mapper profile

Evenya committed:
Fixwd conversions
#bodyazaebal

Evenya committed:
Argument Fix

zefirlover committed:
Merge remote-tracking branch 'origin/STD-35' into STD-40

zefirlover committed:
AddTheoryBlockRequest, RemoveTheoryBlockByIdRequest and TheoryBlockController created

zefirlover committed:
Merge branch 'dev' into STD-40

SleepyF0X committed:
Fix Equals and code style

zefirlover committed:
Merge remote-tracking branch 'origin/STD-35' into STD-40

zefirlover committed:
HttpGet by Id added into TheoryBlockController.cs and other minor changes
Added:
GetTheoryBlockByIdRequest
GetTheoryBlockByIdResponse
GetTheoryBlocksRequest
GetTheoryBlocksResponse
TheoryBlockProfile

WerewolfNet commented on server/StudySharp.API/Controllers/TheoryBlockController.cs:
```diff
@@ -0,0 +1,58 @@
+            _mediator = mediator;
+        }
+
+        [HttpPost("add")]
```
[HttpPost("add")] => [HttpPost]

WerewolfNet commented on server/StudySharp.API/Controllers/TheoryBlockController.cs:
```diff
@@ -0,0 +1,58 @@
+            return await _mediator.Send(addTheoryBlockCommand);
+        }
+
+        [HttpPost("remove")]
```
[HttpPost("remove")] => [HttpDelete("{id:int}")]

WerewolfNet commented on server/StudySharp.API/Controllers/TheoryBlockController.cs:
```diff
@@ -0,0 +1,58 @@
+            return await _mediator.Send(removeTheoryBlockByIdCommand);
+        }
+
+        [HttpGet]
```
[HttpGet] => [HttpGet("{id:int}"]

WerewolfNet commented on server/StudySharp.API/Controllers/TheoryBlockController.cs:
```diff
@@ -0,0 +1,58 @@
+        }
+
+        [HttpGet]
+        public async Task<OperationResult> GetTheoryBlockById([FromBody] GetTheoryBlockByIdRequest getTheoryBlockByIdRequest)
```
`Task<OperationResult>` should be `Task<OperationResult<TheoryBlockResponse>>` OR `Task<OperationResult<TheoryBlock>>`

adjust query and query handler correspondingly

zefirlover replied:
Can i change `OperationResult` to `OperationResult<GetTheoryBlockByIdResponse>`? Rider said that it is correct

WerewolfNet replied:
Респонс должен быть строго типизирован, соответственно в *сигнатуре* метода прошу поменять на один из двух вариантов

WerewolfNet commented on server/StudySharp.API/Requests/TheoryBlocks/GetTheoryBlockByIdRequest.cs:
```diff
@@ -0,0 +1,7 @@
+{
+    public class GetTheoryBlockByIdRequest
+    {
+        public int Id { get; set; }
```
![image](https://user-images.githubusercontent.com/45018119/134784991-ab7a4a92-f768-4296-8f99-889d943295f8.png)

WerewolfNet replied:
Do the same but to GetTheoryBlockByIdRequest accordingly

WerewolfNet commented on server/StudySharp.API/Requests/TheoryBlocks/RemoveTheoryBlockByIdRequest.cs:
```diff
@@ -0,0 +1,7 @@
+{
+    public class RemoveTheoryBlockByIdRequest
+    {
+        public int Id { get; set; }
```
![image](https://user-images.githubusercontent.com/45018119/134785015-9c30a7ba-8c87-4d4b-96c6-cb92b114e7fa.png)

WerewolfNet replied:
Do the same but to RemoveTheoryBlockByIdRequest accordingly

WerewolfNet commented on server/StudySharp.ApplicationServices/Commands/AddCourseCommand.cs:
```diff
@@ -26,19 +26,22 @@ public AddCourseCommandHandler(StudySharpDbContext sharpDbContext)
 
         public async Task<OperationResult> Handle(AddCourseCommand request, CancellationToken cancellationToken)
```
Why did you touch this file?

zefirlover replied:
Mb misclick, i dont remember when and why i changed it

zefirlover replied:
And i dunno what i changed too

WerewolfNet commented on server/StudySharp.API/Controllers/TheoryBlockController.cs:
```diff
@@ -0,0 +1,58 @@
+    [ApiController]
+    [Route("api/[controller]")]
+
+    public class TheoryBlockController : ControllerBase
```
Add Update and Get (all theory blocks for the course by CourseId) methods to the controller

WerewolfNet replied:
GetTheoryBlocksByCourseId is still missing

WerewolfNet commented on server/StudySharp.API/Controllers/TheoryBlockController.cs:
```diff
@@ -0,0 +1,58 @@
+{
+    [Authorize]
+    [ApiController]
+    [Route("api/[controller]")]
```
[Route("api/[controller]")] => [Route("api/theory-blocks")]

zefirlover committed:
After review update

zefirlover committed:
Remove stupid generics

WerewolfNet commented on server/StudySharp.API/Requests/TheoryBlocks/RemoveTheoryBlockByIdRequest.cs:
```diff
@@ -0,0 +1,10 @@
+{
+    public class RemoveTheoryBlockByIdRequest
+    {
+        [BindProperty(Name = "id", SupportsGet = true)]
```
Have you checked the controller functionality with the help of Swagger? Does every method go well? Because here we have SupportsGet flag. Not sure if it will work with HttpDelete, so that's why I'm asking

zefirlover replied:
Not yet

zefirlover committed:
GetTheoryBlockByCourseId pre-update
Add GetTheoryBlockByCourseIdRequest and GetTheoryBlockByCourseIdResponse
Add GetTheoryBlockByCourseId logic to the TheoryBlockController.cs, now commented due to lack of some queries

zefirlover committed:
Add "Update" functionality to the TheoryBlockController

zefirlover committed:
Add GetTheoryBlockByCourseId
Add GetTheoryBlockByCourseIdQuery
Uncomment code in TheoryBlockController.cs

zefirlover committed:
Pre-rewrite update

zefirlover committed:
Merge remote-tracking branch 'origin/STD-38' into STD-40

WerewolfNet commented on server/StudySharp.API/Controllers/TheoryBlockController.cs:
```diff
@@ -0,0 +1,88 @@
+            _mediator = mediator;
+        }
+
+        [HttpPost]
```
[HttpPost] => [HttpPost("{courseId:int}/theory-blocks")]

WerewolfNet commented on server/StudySharp.API/Controllers/TheoryBlockController.cs:
```diff
@@ -0,0 +1,88 @@
+            return await _mediator.Send(addTheoryBlockCommand);
+        }
+
+        [HttpDelete("{id:int}")]
```
[HttpDelete("{id:int}")] => [HttpDelete("{courseId:int}/theory-blocks/{theoryBlockId:int}")]

WerewolfNet commented on server/StudySharp.API/Controllers/TheoryBlockController.cs:
```diff
@@ -0,0 +1,88 @@
+            return await _mediator.Send(removeTheoryBlockByIdCommand);
+        }
+
+        [HttpGet("{id:int}")]
```
[HttpGet("{id:int}")] => [HttpGet("{courseId:int}/theory-blocks/{theoryBlockId:int}")]

WerewolfNet commented on server/StudySharp.API/Controllers/TheoryBlockController.cs:
```diff
@@ -0,0 +1,88 @@
+            return OperationResult.Ok(response);
+        }
+
+        [HttpPut("{id:int}")]
```
[HttpPut("{id:int}")] => [HttpPut("{courseId:int}/theory-blocks/{theoryBlockId:int}")]

WerewolfNet commented on server/StudySharp.API/Controllers/TheoryBlockController.cs:
```diff
@@ -0,0 +1,88 @@
+            return OperationResult.Ok(response);
+        }
+
+        [HttpGet("course/{courseId:int}")]
```
[HttpGet("course/{courseId:int}")] => [HttpGet("{courseId:int}/theory-blocks")]

WerewolfNet commented on server/StudySharp.API/Requests/TheoryBlocks/GetTheoryBlockByCourseIdRequest.cs:
```diff
@@ -0,0 +1,10 @@
+
+namespace StudySharp.API.Requests.TheoryBlocks
+{
+    public class GetTheoryBlockByCourseIdRequest
```
GetTheoryBlockByCourseIdRequest => GetTheoryBlocksByCourseIdRequest

WerewolfNet commented on server/StudySharp.API/Requests/TheoryBlocks/GetTheoryBlocksRequest.cs:
```diff
@@ -0,0 +1,9 @@
+
+namespace StudySharp.API.Requests.TheoryBlocks
+{
+    public class GetTheoryBlocksRequest
```
add CourseId

zefirlover replied:
Lmao
We do not need this file, so i deleted it

WerewolfNet commented on server/StudySharp.API/Requests/TheoryBlocks/RemoveTheoryBlockByIdRequest.cs:
```diff
@@ -0,0 +1,10 @@
+
+namespace StudySharp.API.Requests.TheoryBlocks
+{
+    public class RemoveTheoryBlockByIdRequest
```
add CourseId

WerewolfNet commented on server/StudySharp.API/Requests/TheoryBlocks/UpdateTheoryBlockByIdRequest.cs:
```diff
@@ -0,0 +1,10 @@
+
+namespace StudySharp.API.Requests.TheoryBlocks
+{
+    public class UpdateTheoryBlockByIdRequest
```
add CourseId

WerewolfNet commented on server/StudySharp.API/Responses/TheoryBlocks/GetTheoryBlockByCourseIdResponse.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace StudySharp.API.Responses.TheoryBlocks
+{
+    public class GetTheoryBlockByCourseIdResponse
```
GetTheoryBlockByCourseIdResponse => GetTheoryBlocksByCourseIdResponse

WerewolfNet replied:
Should return list of theory blocks that belong to this course

WerewolfNet commented on server/StudySharp.ApplicationServices/Commands/AddTheoryBlockCommand.cs:
```diff
@@ -0,0 +1,49 @@
+            _context = sharpDbContext;
+        }
+
+        public async Task<OperationResult> Handle(AddTheoryBlockCommand request, CancellationToken cancellationToken)
```
Course existence validation is missing

zefirlover replied:
Umn
So what is that, in the row 31
_.CourseId == request.CourseId

WerewolfNet commented on server/StudySharp.ApplicationServices/Commands/RemoveTheoryBlockByIdCommand.cs:
```diff
@@ -0,0 +1,41 @@
+
+namespace StudySharp.ApplicationServices.Commands
+{
+    public sealed class RemoveTheoryBlockByIdCommand : IRequest<OperationResult>
```
add CourseId

WerewolfNet commented on server/StudySharp.ApplicationServices/Commands/RemoveTheoryBlockByIdCommand.cs:
```diff
@@ -0,0 +1,41 @@
+        public async Task<OperationResult> Handle(
+            RemoveTheoryBlockByIdCommand request,
+            CancellationToken cancellationToken)
+        {
```
Course existence validation is missing

zefirlover replied:
?

WerewolfNet commented on server/StudySharp.ApplicationServices/Commands/UpdateTheoryBlockCommand.cs:
```diff
@@ -0,0 +1,44 @@
+
+namespace StudySharp.ApplicationServices.Commands
+{
+    public sealed class UpdateTheoryBlockCommand : IRequest<OperationResult>
```
add CourseId

WerewolfNet commented on server/StudySharp.ApplicationServices/Commands/UpdateTheoryBlockCommand.cs:
```diff
@@ -0,0 +1,44 @@
+        }
+
+        public async Task<OperationResult> Handle(UpdateTheoryBlockCommand request, CancellationToken cancellationToken)
+        {
```
Course existence validation is missing

WerewolfNet commented on server/StudySharp.ApplicationServices/Queries/GetTheoryBlockByCourseIdQuery.cs:
```diff
@@ -0,0 +1,36 @@
+        }
+
+        public async Task<OperationResult<TheoryBlock>> Handle(GetTheoryBlockByCourseIdQuery request, CancellationToken cancellationToken)
+        {
```
Course existence validation is missing

WerewolfNet commented on server/StudySharp.ApplicationServices/Queries/GetTheoryBlockByCourseIdQuery.cs:
```diff
@@ -0,0 +1,36 @@
+                return OperationResult.Fail<TheoryBlock>(string.Format(ErrorConstants.EntityNotFound, nameof(TheoryBlock), nameof(TheoryBlock.Id), request.CourseId));
+            }
+
+            return OperationResult.Ok(theoryBlock);
```
Should return list of theory blocks that belong to this course

WerewolfNet commented on server/StudySharp.ApplicationServices/Queries/GetTheoryBlockByIdQuery.cs:
```diff
@@ -0,0 +1,38 @@
+
+namespace StudySharp.ApplicationServices.Queries
+{
+    public sealed class GetTheoryBlockByIdQuery : IRequest<OperationResult<TheoryBlock>>
```
add CourseId

WerewolfNet commented on server/StudySharp.ApplicationServices/Queries/GetTheoryBlockByIdQuery.cs:
```diff
@@ -0,0 +1,38 @@
+        public async Task<OperationResult<TheoryBlock>> Handle(
+            GetTheoryBlockByIdQuery request,
+            CancellationToken cancellationToken)
+        {
```
Course existence validation is missing

WerewolfNet commented on server/StudySharp.API/Responses/TheoryBlocks/UpdateTheoryBlockResponce.cs:
```diff
@@ -0,0 +1,6 @@
+namespace StudySharp.API.Responses.TheoryBlocks
+{
+    public class UpdateTheoryBlockResponce
```
UpdateTheoryBlockResponce => UpdateTheoryBlockResponse

zefirlover committed:
Major update of the TheoryBlockController
TheoryBlockController now works like a part of Course
Remove GetTheoryBlocksRequest.cs due to is uselessness
Other minor changes due to adding CourseId validation

zefirlover committed:
GetTheoryBlockByCourseIdQuery returns a List of TheoryBlocks if theoryBlock == null

zefirlover committed:
Merge branch 'dev' into STD-40

# Conflicts:
#	server/StudySharp.ApplicationServices/Commands/RemoveTheoryBlockByIdCommand.cs
#	server/StudySharp.ApplicationServices/Queries/GetTheoryBlockByIdQuery.cs

zefirlover committed:
Pull update from dev

zefirlover committed:
Add Course validation
In AddTheoryBlockCommand, GetTheoryBlockByCourseIdQuery, GetTheoryBlockByIdQuery, RemoveTheoryBlockByIdCommand, UpdateTheoryBlockCommand

zefirlover committed:
Remove useless comment

WerewolfNet commented on server/StudySharp.API/Controllers/CourseController.cs:
```diff
@@ -0,0 +1,73 @@
+    [ApiController]
+    [Route("api/[controller]")]
+
+    public class CourseController : ControllerBase
```
Delete this file

WerewolfNet commented on server/StudySharp.API/MapperProfiles/CourseProfile.cs:
```diff
@@ -0,0 +1,18 @@
+{
+    public class CourseProfile : Profile
+    {
+        public CourseProfile()
```
delete this file

WerewolfNet commented on server/StudySharp.API/Requests/Courses/AddCourseRequest.cs:
```diff
@@ -0,0 +1,8 @@
+﻿namespace StudySharp.API.Requests.Courses
+{
+    public class AddCourseRequest
```
delete this file

WerewolfNet commented on server/StudySharp.API/Requests/Courses/GetCourseByIdRequest.cs:
```diff
@@ -0,0 +1,7 @@
+﻿namespace StudySharp.API.Requests.Courses
+{
+    public class GetCourseByIdRequest
```
delete this file

WerewolfNet commented on server/StudySharp.API/Requests/Courses/GetCoursesRequest.cs:
```diff
@@ -0,0 +1,6 @@
+﻿namespace StudySharp.API.Requests.Courses
+{
+    public class GetCoursesRequest
```
delete this file

WerewolfNet commented on server/StudySharp.API/Requests/Courses/RemoveCourseByIdRequest.cs:
```diff
@@ -0,0 +1,7 @@
+﻿namespace StudySharp.API.Requests.Courses
+{
+    public class RemoveCourseByIdRequest
```
delete this file

WerewolfNet commented on server/StudySharp.API/Responses/Courses/AddCourseResponse.cs:
```diff
@@ -0,0 +1,6 @@
+﻿namespace StudySharp.API.Responses.Courses
+{
+    public class AddCourseResponse
```
delete this file

WerewolfNet commented on server/StudySharp.API/Responses/Courses/GetCourseByIdResponse.cs:
```diff
@@ -0,0 +1,18 @@
+using StudySharp.Domain.Models;
+namespace StudySharp.API.Responses.Courses
+{
+    public class GetCourseByIdResponse
```
delete this file

WerewolfNet commented on server/StudySharp.API/Responses/Courses/GetCoursesResponse.cs:
```diff
@@ -0,0 +1,10 @@
+
+namespace StudySharp.API.Responses.Courses
+{
+    public class GetCoursesResponse
```
delete this file

WerewolfNet commented on server/StudySharp.API/Responses/Courses/RemoveCourseByIdResponse.cs:
```diff
@@ -0,0 +1,6 @@
+﻿namespace StudySharp.API.Responses.Courses
+{
+    public class RemoveCourseByIdResponse
```
delete this file

WerewolfNet commented on server/StudySharp.API/Responses/TheoryBlocks/GetTheoryBlocksResponse.cs:
```diff
@@ -0,0 +1,10 @@
+
+namespace StudySharp.API.Responses.TheoryBlocks
+{
+    public class GetTheoryBlocksResponse
```
Theory blocks can only exist in the scope of course. This file is redundant. Delete this file

WerewolfNet commented on server/StudySharp.API/Controllers/TheoryBlockController.cs:
```diff
@@ -0,0 +1,88 @@
+        [HttpPut("{courseId:int}/theory-blocks/{theoryBlockId:int}")]
+        public async Task<OperationResult<UpdateTheoryBlockResponse>> Update([FromRoute] UpdateTheoryBlockByIdRequest updateTheoryBlockByIdRequest)
+        {
+            var updateTheoryBlockCommand = _mapper.Map<UpdateTheoryBlockCommand>(updateTheoryBlockByIdRequest);
```
From where do you get updated Name value for the theory block? It is missing in the request model. Have you even tested it locally?

zefirlover replied:
I do not, i have another problem with local database

WerewolfNet commented on server/StudySharp.API/Controllers/TheoryBlockController.cs:
```diff
@@ -0,0 +1,88 @@
+        }
+
+        [HttpGet("{courseId:int}/theory-blocks")]
+        public async Task<OperationResult<GetTheoryBlocksByCourseIdResponse>> GetTheoryBlockByCourseId([FromRoute] GetTheoryBlocksByCourseIdRequest getTheoryBlockByCourseIdRequest)
```
GetTheoryBlockByCourseId => GetTheoryBlocksByCourseId

WerewolfNet commented on server/StudySharp.API/Requests/TheoryBlocks/UpdateTheoryBlockByIdRequest.cs:
```diff
@@ -0,0 +1,11 @@
+
+namespace StudySharp.API.Requests.TheoryBlocks
+{
+    public class UpdateTheoryBlockByIdRequest
```
field Name is missing

WerewolfNet replied:
TheoryBlockId and CourseId should be grabbed from Route, Name should be grabbed from Body

zefirlover committed:
Minor changes
Delete GetTheoryBlocksResponse.cs
GetTheoryBlockByCourseId => GetTheoryBlocksByCourseId

zefirlover committed:
Remove unnecessary Course elements

zefirlover committed:
Tests 1/2
Add row in mapper profile
Rename and minor changes in GetTheoryBlocksByCourseIdQuery.cs

zefirlover committed:
Tests 2/2
Some changes in Update and Remove blocks

zefirlover committed:
Tests 3/2
Fixed CourseId validation in UpdateTheoryBlockCommand.cs
Update now works correctly
Other minor changes

zefirlover committed:
Add now works, rewrite operation result text in GetTheoryBlockByIdQuery
Other minor changes

zefirlover committed:
Rework courseId exception in Remove, GetById, GetByCourseId, rework exception texts in GetByCourseId

zefirlover committed:
Unnecessary comments was removed

zefirlover committed:
Change course to courseExistent
